sdks/python/apache_beam/transforms/util.py

```diff
@@ -161,3 +174,215 @@ def RemoveDuplicates(pcoll):  # pylint: disable=invalid-name
           | 'ToPairs' >> Map(lambda v: (v, None))
           | 'Group' >> CombinePerKey(lambda vs: None)
           | 'RemoveDuplicates' >> Keys())
+
+
+class _BatchSizeEstimator(object):
+  """Estimates the best size for batches given historical timing.
+  """
+
+  MAX_DATA_POINTS = 100
+  MAX_GROWTH_FACTOR = 2
+
+  def __init__(self,
+               min_batch_size=1,
+               max_batch_size=1000,
+               target_batch_overhead=.1,
+               target_batch_duration=1,
+               clock=time.time):
+    if min_batch_size > max_batch_size:
+      raise ValueError("Minimum (%s) must not be greater than maximum (%s)" % (
+          min_batch_size, max_batch_size))
+    self._min_batch_size = min_batch_size
+    self._max_batch_size = max_batch_size
+    self._target_batch_overhead = target_batch_overhead
+    self._target_batch_duration = target_batch_duration
+    self._clock = clock
+    self._data = []
+    self._ignore_next_timing = False
+    self._size_distribution = Metrics.distribution(
+        'BatchElements', 'batch_size')
+    self._time_distribution = Metrics.distribution(
+        'BatchElements', 'msec_per_batch')
+    self._remainder_msecs = 0
+
+  def ignore_next_timing(self, ignore_next_timing=True):
```

Comment:
<!--thread_id:cc_144437524_t; commit:85e7756335b0cf7501f62bd283b07fa9ea468a21; resolved:1-->
<!--section:context-quote-->
> **katsiapis** wrote:
> Why is the argument needed? Do we ever call it with False?

<!--section:body-->
It seemed a more complete API to allow unsetting rather than just setting. However, it's a backwards compatible change to add it later if needed, so removed for now.